File: inst/include/cpp11_methods/sample.cpp

[[cpp11::register]]
void sample(cpp11::list args_list) {
  model_ptr_t model(args_list["model_ptr"]);
  std::string rdump_init = cpp11::as_cpp<std::string>(args_list["rdump_init"]);
  std::string output = cpp11::as_cpp<std::string>(args_list["output"]);
  size_t num_chains = cpp11::as_cpp<size_t>(args_list["num_chains"]);
  size_t num_threads = cpp11::as_cpp<size_t>(args_list["num_threads"]);

  std::shared_ptr<stan::io::var_context> init_ptr = var_context(rdump_init);
  std::vector<std::shared_ptr<stan::io::var_context>> init_contexts;
  for (int i = 0; i < num_chains; ++i) {
    init_contexts.push_back(init_ptr);
  }

  stan::math::init_threadpool_tbb(num_threads);

  R_CheckUserInterrupt_Functor interrupt;
  stan::callbacks::stream_logger logger(std::cout, std::cout, std::cout, std::cerr, std::cerr);

  std::vector<stan::callbacks::unique_stream_writer<std::ostream>>
      sample_writers;
  sample_writers.reserve(num_chains);
  std::vector<stan::callbacks::unique_stream_writer<std::ostream>>
      diagnostic_writers;
  diagnostic_writers.reserve(num_chains);
  std::vector<stan::callbacks::writer> init_writers{num_chains,
                                                    stan::callbacks::writer{}};

  for (int i = 0; i < num_chains; ++i) {
    auto output_filename = output + "_" + std::to_string(i) + ".csv";
    auto unique_fstream
        = std::make_unique<std::fstream>(output_filename, std::fstream::out);

    sample_writers.emplace_back(std::move(unique_fstream), "# ");
    diagnostic_writers.emplace_back(nullptr, "# ");
  }

  int return_code = stan::services::sample::hmc_nuts_dense_e_adapt(
    *model.get(),
    num_chains,
    init_contexts,
    cpp11::as_cpp<size_t>(args_list["random_seed"]),
    cpp11::as_cpp<int>(args_list["id"]),
    cpp11::as_cpp<double>(args_list["init_radius"]),
    cpp11::as_cpp<int>(args_list["num_warmup"]),
    cpp11::as_cpp<int>(args_list["num_samples"]),
    cpp11::as_cpp<int>(args_list["num_thin"]),
    cpp11::as_cpp<bool>(args_list["save_warmup"]),
    cpp11::as_cpp<int>(args_list["refresh"]),
    cpp11::as_cpp<double>(args_list["stepsize"]),
    cpp11::as_cpp<double>(args_list["stepsize_jitter"]),
    cpp11::as_cpp<int>(args_list["max_depth"]),
    cpp11::as_cpp<double>(args_list["delta"]),
    cpp11::as_cpp<double>(args_list["gamma"]),
    cpp11::as_cpp<double>(args_list["kappa"]),
    cpp11::as_cpp<double>(args_list["t0"]),
    cpp11::as_cpp<size_t>(args_list["init_buffer"]),
    cpp11::as_cpp<size_t>(args_list["term_buffer"]),
    cpp11::as_cpp<size_t>(args_list["window"]),
    interrupt,
    logger,
    init_writers,
    sample_writers,
    diagnostic_writers);
}

File: R/sample.R

sample <- function(
  data,
  init = NULL,
  output_dir = tempdir(),
  num_chains = 4,
  num_threads = 1,
  random_seed = 0,
  num_samples = 2000,
  num_warmup = 2000,
  save_warmup = FALSE,
  num_thin = 1,
  adapt = TRUE,
  gamma = 0.05,
  delta = 0.8,
  kappa = 0.75,
  t0 = 10,
  init_buffer = 75,
  term_buffer = 50,
  window = 25,
  algorithm = "hmc",
  engine = "nuts",
  max_depth = 10,
  metric = "diag_e",
  inv_metric = NULL,
  stepsize = 1,
  stepsize_jitter = 0,
  init_radius = 2,
  refresh = 200
  ) {

  input_args <- as.list(environment())

  service_name <- algorithm

  if (algorithm != "fixed_param") {
    service_name <- paste0(c(service_name, engine, metric), collapse = "_")
    if (adapt) {
      service_name <- paste0(c(service_name, "adapt"), collapse = "_")
    }
  }
  if (!is.null(init)) {
    input_args$rdump_init <- stan_rdump(init)
  }
  input_args$rdump_data <- stan_rdump(data)
  input_args$id <- 1
  if (!is.null(inv_metric)) {
    if (metric == "diag_e") {
      input_args$inv_metric <- paste0(
        "inv_metric <- c(",
        paste(diag(inv_metric), collapse = ","), ")"
      )
    } else {
      input_args$inv_metric <- stan_rdump(list(inv_metric=inv_metric))
    }
  }

  private$env$stan_methods_wrapper(service_name, input_args)
  model_ptr <- private$env$new_model(input_args$rdump_data, seed = random_seed)
  outputs <- paste0(input_args$output_dir, "/output", "_", 1:(num_chains), ".csv")
  stanfit$new(self, private, outputs, model_ptr)
}
